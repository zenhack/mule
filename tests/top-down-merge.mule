#XFAIL?
#
# The type inference algorithm requires merging types in a "top down"
# manner. If you do bottom up instead (which is eaiser), you get something
# that is sound, but incomplete. This test checks that merges are being
# done top-down, as required.
{ type t a = {} } : exist b. { type t a = b }
