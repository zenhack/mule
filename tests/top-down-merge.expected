
Desugared:

	((: (exist b (record (types ((t (lam a (rec t b))))) (values ()))))
	 (fix/record
	  (fn $0
	   (let-type ((t (lam a (rec t (record (types ()) (values ()))))))
	    (((.type= t) {}) (type t))))))

inferred type:

	(exist a (record (types ((t (lam b a)))) (values ())))

Runtime term:

	(App ((Fix Record) (Lam (0 () (Record ())))))

Evaluated:

	(Record ())
